File: application/src/main/java/com/example/server/service/impl/ItemServiceImpl.java
package com.example.server.service.impl;

import com.example.api.dto.ItemDto;
import com.example.api.dto.ItemDtoWithBookings;
import com.example.server.repository.FromSizePageRequest;
import com.example.server.repository.ItemRepository;
import com.example.server.repository.BookingRepository;
import com.example.server.repository.ItemRequestRepository;
import com.example.server.repository.UserRepository;
import com.example.server.repository.entity.Booking;
import com.example.server.repository.CommentRepository;
import com.example.server.repository.entity.Comment;
import com.example.server.exception.NotFoundException;
import com.example.server.mapper.ItemMapper;
import com.example.server.repository.entity.Item;
import com.example.server.repository.entity.ItemRequest;
import com.example.server.repository.entity.User;
import com.example.server.service.ItemService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;

@RequiredArgsConstructor
@Service
public class ItemServiceImpl implements ItemService {

    private final BookingRepository bookingRepository;
    private final CommentRepository commentRepository;
    private final ItemMapper itemMapper;
    private final ItemRepository itemRepository;
    private final ItemRequestRepository itemRequestRepository;
    private final UserRepository userRepository;

    @Override
    @Transactional
    public Item create(ItemDto itemDto, Long userId) {
        User owner = userRepository.findById(userId)
                .orElseThrow(() -> new NotFoundException("Пользователь с указанным идентификатором не найден"));
        ItemRequest itemRequest = itemDto.getRequestId() == null ? null : itemRequestRepository.findById(itemDto.getRequestId())
                .orElseThrow(() -> new NotFoundException("Запрос на добавление предмета с указанным идентификатором не найден"));

        return itemRepository.save(itemMapper.toItem(itemDto, owner, itemRequest));
    }

    @Override
    @Transactional
    public Item update(Long id, ItemDto itemDto, Long userId) {
        Item item = findById(id);
        Long ownerId = item.getOwner().getId();

        if (!ownerId.equals(userId)) {
            throw new NotFoundException("Обновлять информацию о предмете может только его владелец");
        }
        return itemRepository.save(itemMapper.updateItem(itemDto, item));
    }

    @Override
    public Item findById(Long id) {
        return itemRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Предмет с указанным идентификатором не найден"));
    }

    @Override
    public ItemDtoWithBookings findByIdWithBooking(Long id, Long userId) {
        Item item = findById(id);
        Long ownerId = item.getOwner().getId();

        if (ownerId.equals(userId)) {
            return itemMapper.toItemDtoWithBookings(item, bookingRepository.findAllByItemId(id), commentRepository.findAllByItemId(id));
        }
        return itemMapper.toItemDtoWithBookings(item, commentRepository.findAllByItemId(id));
    }

    @Override
    public List<ItemDtoWithBookings> findAllByOwnerId(Long userId, Integer from, Integer size) {
        List<Item> items = itemRepository.findAllByOwnerId(userId, FromSizePageRequest.of(from, size, Sort.by(Sort.Direction.ASC, Item.Fields.id)));
        List<Booking> bookings = bookingRepository.findAllByItem_Owner_Id(userId, Pageable.unpaged());
        List<Comment> comments = commentRepository.findAllByOwnerId(userId);

        return itemMapper.toItemDtoWithBookings(items, bookings, comments);
    }

    @Override
    public List<Item> search(String text, Integer from, Integer size) {
        if (text.isBlank()) {
            return Collections.emptyList();
        }
        return itemRepository.search(text.toLowerCase(), FromSizePageRequest.of(from, size));
    }
}

File: application/src/main/kotlin/com/example/server/service/impl/ItemServiceImpl.kt
package com.example.server.service.impl

import com.example.api.model.ItemDto
import com.example.api.model.ItemDtoWithBooking
import com.example.server.entity.Item
import com.example.server.exception.NotFoundException
import com.example.server.mapper.ItemMapper
import com.example.server.repository.BookingRepository
import com.example.server.repository.CommentRepository
import com.example.server.repository.FromSizePageRequest.Companion.of
import com.example.server.repository.ItemRepository
import com.example.server.repository.ItemRequestRepository
import com.example.server.repository.UserRepository
import com.example.server.service.ItemService
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ItemServiceImpl(
    private val bookingRepository: BookingRepository,
    private val commentRepository: CommentRepository,
    private val itemMapper: ItemMapper,
    private val itemRepository: ItemRepository,
    private val itemRequestRepository: ItemRequestRepository,
    private val userRepository: UserRepository
) : ItemService {

    @Transactional
    override fun create(itemDto: ItemDto, userId: Long): Item {
        val owner = userRepository.findById(userId)
            .orElseThrow { NotFoundException(UserRepository.NOT_FOUND) }
        val itemRequest = if (itemDto.requestId == null) null else itemRequestRepository.findById(itemDto.requestId!!)
            .orElseThrow { NotFoundException(ItemRequestRepository.NOT_FOUND) }

        return itemRepository.save(itemMapper.toEntity(itemDto, owner, itemRequest))
    }

    @Transactional
    override fun update(id: Long, itemDto: ItemDto, userId: Long): Item {
        val item = findById(id)

        if (userId != item.owner.id) {
            throw NotFoundException("Обновлять информацию о предмете может только его владелец")
        }
        return itemMapper.updateEntity(itemDto, item)
    }

    override fun findById(id: Long): Item {
        return itemRepository.findById(id)
            .orElseThrow { NotFoundException(ItemRepository.NOT_FOUND) }
    }

    @Transactional(readOnly = true)
    override fun findByIdWithBooking(id: Long, userId: Long): ItemDtoWithBooking {
        val item = findById(id)

        if (userId != item.owner.id) {
            return itemMapper.toDtoWithBooking(item, emptyList(), commentRepository.findAllByItemId(id))
        }
        return itemMapper.toDtoWithBooking(item, bookingRepository.findAllByItemId(id), commentRepository.findAllByItemId(id))
    }

    @Transactional(readOnly = true)
    override fun findAllByOwnerId(userId: Long, from: Int, size: Int): List<ItemDtoWithBooking> {
        val pageable = of(from, size, SORT_BY_DESCENDING_ID)
        val items = itemRepository.findAllByOwnerId(userId, pageable)
        val bookings = bookingRepository.findAllByItem_Owner_Id(userId, Pageable.unpaged())
        val comments = commentRepository.findAllByOwnerId(userId)

        return itemMapper.toDtoWithBooking(items, bookings, comments)
    }

    override fun search(text: String, from: Int, size: Int): List<Item> {
        return if (text.isBlank()) emptyList() else itemRepository.search(text.lowercase(), of(from, size))
    }

    companion object {
        private val SORT_BY_DESCENDING_ID = Sort.by(Sort.Direction.ASC, "id")
    }
}
